### ISSUE TITLE
Work more explanation into the earlier parts of 'Twisted from Scratch'

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11539]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)| @puzzlement reported|
|-|-|
|Trac ID|trac#1025|
|Type|enhancement|
|Created|2005-05-08 17:28:46Z|
|Branch|https://github.com/twisted/twisted/tree/twisted-scratch-1025|

<details><summary>Searchable metadata</summary>

```
trac-id__1025 1025
type__enhancement enhancement
reporter__hypatia hypatia
priority__high high
milestone__ 
branch__branches_twisted_scratch_1025 branches/twisted-scratch-1025
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1115573326000000 1115573326000000
changetime__1220227370000000 1220227370000000
version__ 
owner__ 
cc__Predictive cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement commented|
|-|-|

```
#!html
<pre>
Suggestions from #1015:

- Twisted from Scratch gets confusing early on because so much is presented in
the first three sections. I would rework it so:

- In the very first page as you explain what each line does, explain that

  - A reactor is not magical and scary, it's just the event loop like
    Mainloop() in Tk. It selects (polls, kqueues) so you don't have to.

  - A Protocol defines what you want to do in response to those events, and
    you get a new one for every connection.

  - A Factory spits out new Protocol instances (more or less), and is the 
    proper place for data that you want all the protocol instances to have 
    access to, since they die when the connection is closed.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement set owner to edsuom|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra commented|
|-|-|

(In [24553]) Branching to 'twisted-scratch-1025'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Added Predictive's notes, putting the ticket up for review.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)|@washort set owner to @washort<br>@washort set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_8"></a>|@washort set owner to @thijstriemstra<br>@washort set status to `new`|
|-|-|

Looks OK; merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

The new trailing whitespace in the paragraph which begins _Here, we start listening on port 1079._ may not be much of an improvement.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_9):
> The new trailing whitespace in the paragraph which begins _Here, we start listening on port 1079._ may not be much of an improvement.

I removed some whitespace in r24612 but not sure which one you meant, so putting it up for review again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [24617]) remove trailing whitespace

refs [#11539](https://github.com/twisted/twisted/issues/11539)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

That was the whitespace I meant. :)

If you use emacs, `M-x customize-variable show-trailing-whitespace`, turning this on makes it obvious where whitespace like this exists.

Regarding the rest of the branch, I didn't want to comment before, but as long as the ticket is assigned to me for review now...

It seems likely that Tk is the most obscure toolkit we could make reference to here.  I suspect more readers will be familiar with any of the other toolkits available in Python (Qt, wx, Gtk).  The mainloop API in `Tkinter` is also `Tkinter.Tk.mainloop`, not `Mainloop` (perhaps it is `Mainloop` in the TCL API, I don't know).

I'd also be happier with slightly more precise/formal language in the entirely new paragraph - rather than _spits_, _instantiates_ or _creates_, even _produces_.  _are discarded_ or _are garbage collected_ instead of _die_.

Thanks

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra removed owner|
|-|-|

Replying to [exarkun](#note_12):
> That was the whitespace I meant. :)

Thanks :)

> It seems likely that Tk is the most obscure toolkit we could make reference to here.  I suspect more readers will be familiar with any of the other toolkits available in Python (Qt, wx, Gtk).  The mainloop API in `Tkinter` is also `Tkinter.Tk.mainloop`, not `Mainloop` (perhaps it is `Mainloop` in the TCL API, I don't know).
>
> I'd also be happier with slightly more precise/formal language in the entirely new paragraph - rather than _spits_, _instantiates_ or _creates_, even _produces_.  _are discarded_ or _are garbage collected_ instead of _die_.

Rewrote it in r24637, putting it up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_15"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [24653]) Merge twisted-scratch-1025: Adding more explanation into the earlier parts of the 'Twisted from Scratch' doc.

Author: thijs
Reviewer: exarkun
Fixes: [#11539](https://github.com/twisted/twisted/issues/11539)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
